src/core/lib/channel/channel_tracer.cc

```diff
@@ -0,0 +1,275 @@
+      TraceEvent* to_free = it;
+      it = it->next_;
+      FreeNode(to_free);
+    }
+    gpr_mu_destroy(&tracer_mu_);
+  }
+}
+
+intptr_t ChannelTracer::GetUuid() { return channel_uuid_; }
+
+void ChannelTracer::AddTrace(grpc_slice data, grpc_error* error,
+                             grpc_connectivity_state connectivity_state,
+                             ChannelTracer* referenced_tracer) {
+  if (!max_list_size_) return;  // tracing is disabled if max_nodes == 0
+  ++num_nodes_logged_;
+  // create and fill up the new node
+  TraceEvent* new_trace_node =
+      New<TraceEvent>(data, error, connectivity_state, referenced_tracer);
+  // first node case
+  if (head_trace_ == nullptr) {
+    head_trace_ = tail_trace_ = new_trace_node;
+  }
+  // regular node add case
+  else {
+    tail_trace_->next_ = new_trace_node;
+    tail_trace_ = tail_trace_->next_;
+  }
+  ++list_size_;
+  // maybe garbage collect the end
+  if (list_size_ > max_list_size_) {
+    TraceEvent* to_free = head_trace_;
+    head_trace_ = head_trace_->next_;
+    FreeNode(to_free);
+    --list_size_;
+  }
+}
+
+// returns an allocated string that represents tm according to RFC-3339.
+static char* fmt_time(gpr_timespec tm) {
+  char buffer[35];
+  struct tm* tm_info = localtime((const time_t*)&tm.tv_sec);
+  strftime(buffer, sizeof(buffer), "%Y-%m-%dT%H:%M:%S", tm_info);
+  char* full_time_str;
+  gpr_asprintf(&full_time_str, "%s.%09dZ", buffer, tm.tv_nsec);
+  return full_time_str;
+}
+
+class ChannelTracerRenderer {
+ public:
+  ChannelTracerRenderer(ChannelTracer* tracer, bool recursive)
+      : current_tracer_(tracer),
+        recursive_(recursive),
+        seen_tracers_(nullptr),
+        size_(0),
+        cap_(0) {}
+
+  char* Run() {
+    grpc_json* json = grpc_json_create(GRPC_JSON_OBJECT);
+    AddSeenTracer(current_tracer_);
+    RecursivelyPopulateJson(json);
+    gpr_free(seen_tracers_);
+    char* json_str = grpc_json_dump_to_string(json, 1);
+    grpc_json_destroy(json);
+    return json_str;
+  }
+
+ private:
+  void AddSeenTracer(ChannelTracer* newly_seen) {
+    if (size_ >= cap_) {
+      cap_ = GPR_MAX(5 * sizeof(newly_seen), 3 * cap_ / 2);
+      seen_tracers_ = (ChannelTracer**)gpr_realloc(seen_tracers_, cap_);
+    }
+    seen_tracers_[size_++] = newly_seen;
+  }
+
+  bool TracerAlreadySeen(ChannelTracer* tracer) {
+    for (size_t i = 0; i < size_; ++i) {
+      if (seen_tracers_[i] == tracer) return true;
+    }
+    return false;
+  }
+
+  void RecursivelyPopulateJson(grpc_json* json) {
+    grpc_json* channel_data = grpc_json_create_child(
+        nullptr, json, "channelData", nullptr, GRPC_JSON_OBJECT, false);
+    grpc_json* children = nullptr;
+    if (recursive_) {
+      children = grpc_json_create_child(channel_data, json, "children", nullptr,
+                                        GRPC_JSON_ARRAY, false);
+    }
+    PopulateTracer(channel_data, children);
+  }
+
+  void PopulateTracer(grpc_json* channel_data, grpc_json* children) {
```

Comment:
need some documentation here.

Reply:
Looks functionally correct. not sure if this part is just for testing, but if it's more permanent, then variables need a bit of remaining or just more documentation. child, children, nodes are a bit confusing :D

Reply:
I took a stab at adding more comments and renaming a few variables. Let me know if you think it is sufficient